src/connectors/impls/gcl/writer/meta.rs

```diff
@@ -0,0 +1,604 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::errors::Result;
+use beef::Cow;
+use googapis::google::logging::{
+    r#type::{HttpRequest, LogSeverity},
+    v2::{LogEntryOperation, LogEntrySourceLocation},
+};
+use halfbrown::HashMap;
+use std::collections::HashMap as StdHashMap;
+use tremor_value::Value;
+use value_trait::ValueAccess;
+
+use super::Config;
+
+pub(crate) fn default_log_name(config: &Config) -> String {
+    // Use user supplied default, if provided
+    if let Some(has_usersupplied_default) = &config.log_name {
+        return has_usersupplied_default.clone();
+    }
+
+    // Use hardwired `default`, if no user supplied default provided
+    "default".to_string()
+}
+
+pub(crate) fn log_name(config: &Config, meta: Option<&HashMap<Cow<str>, Value>>) -> String {
+    // Override for a specific per event log_name
+    if let Some(has_meta) = meta {
+        if let Some(log_name) = has_meta.get("log_name") {
+            return log_name.to_string();
+        }
+    }
+
+    default_log_name(config)
+}
+
+pub(crate) fn log_severity(
+    config: &Config,
+    meta: Option<&HashMap<Cow<str>, Value>>,
+) -> Result<i32> {
+    // Override for a specific per event log_severity
+    if let Some(has_meta) = meta {
+        if let Some(log_severity) = has_meta.get("log_severity") {
+            #[allow(clippy::redundant_else)] // Not redundant
+            if let Some(log_severity) = log_severity.as_i32() {
+                return Ok(log_severity);
+            } else {
+                return Err(
+                    "Unable to parse `log_severity` from event metadata - expected i32".into(),
+                );
+            }
+        }
+    }
+
+    // No per event override, use user supplied default
+    if let Some(default_severity) = config.default_severity {
+        return Ok(default_severity);
+    }
+
+    // No per event or user supplied configuration default, use underlying default
+    Ok(LogSeverity::Default as i32)
+}
+
+pub(crate) fn insert_id(meta: Option<&HashMap<Cow<str>, Value>>) -> String {
+    // Override for a specific per event log_severity
+    if let Some(has_meta) = meta {
+        if let Some(insert_id) = has_meta.get("insert_id") {
+            return insert_id.to_string();
+        }
+    }
+
+    // NOTE user supplied defaults supported at this time
+
+    "".to_string() // Allow GCL to auto assign a unique id for this event
+}
+
+pub(crate) fn http_request(meta: Option<&HashMap<Cow<str>, Value>>) -> Option<HttpRequest> {
+    // Override for a specific per event trace
+    if let Some(has_meta) = meta {
```

Comment:
This can be un nested by using:

```rust
fn ... {
  let meta = meta?
  let http_request = has_meta.get("http_request")?;

  Soime(...)
}
```

that would make the function a lot more readable